### PR TITLE
refactor the conflict msg to chef_object behaviour

### PR DESCRIPTION
**Pull request by sdelano**
See: https://github.com/opscode/chef_objects/pull/56

/cc @seth @oferrigni

### CODE REVIEW

File: src/chef_wm_base.erl [hunk 1]
```diff
@@ -432,26 +432,6 @@ assemble_principal_ejson(#principal_state{name = Name,
       {<<"type">>, Type},
       {<<"authz_id">>, AuthzId}]}.
 
-conflict_message(cookbook_version, _Name) ->
```
Comment (seth) [inline src/chef_wm_base.erl:435]: ok, I _do_ like what the refactor does in terms of cleaning up this hydra-fun :)
still would like to discuss some.

**Comment by sdelano** [comment]
@seth Refactored now to use the resource modules instead of the `chef_object` behaviour.

File: src/chef_wm_data.erl [hunk 2]
```diff
@@ -103,3 +103,8 @@ from_json(Req, #base_state{resource_state =
 
 malformed_request_message(Any, _Req, _State) ->
     error({unexpected_malformed_request_message, Any}).
+
+-spec conflict_message(binary()) -> ejson_term().
+conflict_message(_Name) ->
+    %% {[{<<"error">>, [<<"Data Bag '", Name/binary, "' already exists">>]}]}.
```
Comment (seth) [inline src/chef_wm_data.erl]: Let's remove the comment code. I know this was there before, but seems silly to keep carrying it around.

**Comment by seth** [comment]
This looks good and I like the organization better than the initial pass (where stuff moved into chef_objects).

Was it intentional to not add the callback to `chef_wm.erl`?

**Comment by sdelano** [comment]
@seth yes it was intentional not to make this a `chef_wm` callback and I'm not sure which is the right choice. There are a lot of modules that will never call `create_from_json` or `update_from_json` and hence will not need this code. If we leave it as implemented, we'll get an undef if we missed a module. If we make this a `chef_wm` callback, we'll simply have to throw `not_implemented` for all of the modules that won't use this. I'm not sold either way here.

**Comment by seth** [comment]
@sdelano I'm fine with this merging as-is. I agree that with the current organization it's a toss up whether there's any benefit to putting a new callback in the chef_wm behavior.

Looking at the collection of modules, however, it might not be unreasonable to introduce a new behavior. Consider the following.

These modules call `create_from_json` or `update_from_json` and implement the `chef_wm` behavior:

```
src/chef_wm_clients.erl
src/chef_wm_cookbook_version.erl
src/chef_wm_data.erl
src/chef_wm_environments.erl
src/chef_wm_named_client.erl
src/chef_wm_named_data_item.erl
src/chef_wm_named_environment.erl
src/chef_wm_named_node.erl
src/chef_wm_named_role.erl
src/chef_wm_named_user.erl
src/chef_wm_nodes.erl
src/chef_wm_roles.erl
src/chef_wm_users.erl
```

And here are the modules that do `chef_wm` but don't call create or update from json:

```
src/chef_wm_authenticate_user.erl
src/chef_wm_cookbooks.erl
src/chef_wm_depsolver.erl
src/chef_wm_environment_cookbooks.erl
src/chef_wm_environment_recipes.erl
src/chef_wm_environment_roles.erl
src/chef_wm_named_data.erl
src/chef_wm_named_principal.erl
src/chef_wm_named_sandbox.erl
src/chef_wm_sandboxes.erl
src/chef_wm_search.erl
src/chef_wm_search_index.erl
```

The first group seems to represent common Chef object resource patterns and the second group represent the odd balls.

**Comment by sdelano** [comment]
I was thinking along the same lines for the "common" chef object resource patterns. Let's chat about this sometime tomorrow before I head out for the week. I think, design-wise, it's a good middle ground between the one extreme here of only implementing the callbacks you know you need (like is done in this PR) and forcing all callbacks to be implemented.

However, I think the benefits of such structure are mostly realized while in the development process of creating new resources. A common object resource behaviour is a solid framework from which the build (or port) new resources on top of. If such a framework existed for only the "common" Chef objects, then any new objects added would most likely fall outside of the "common" requirement and we wouldn't necessarily realize these benefits.

**Comment by marcparadise** [comment]
conflict message and others have (see chef_wm_util vs chef_wm_malformed) gotten worse, this is probably worth a fresh look under current structure under oc_erchef

**Pull request closed without merging**